### PR TITLE
197 session timeout

### PR DESCRIPTION
**Pull request by carltonsmith**
### This pull request changes...

This makes it so that a user will be logged out after being inactive. The system will update the session on both the frontend and backend when a user interacts with the system (backend updates are limited to 1 per second for testing purposes, but will be bumped up to 30 seconds before merge). 

When the user stops interacting with the system for 6 seconds (updating to 20 minutes before merge) a modal window appears asking if the user would like to extend their session or log out of the system. Failure to respond will result in the user automatically being logged out after 4 additional seconds (updating to 3 minutes before merge).

The backend will timeout the session after 5 minutes (updating to 30 minutes before merge). The user should never experience the backend timing out

It closes raft issues [#197](https://github.com/raft-tech/TANF-app/issues/197) and [#278](https://github.com/raft-tech/TANF-app/issues/278), as well as a11y issue [#425](https://github.com/raft-tech/TANF-app/issues/425)

### Deliverable 1: Accepted Features

> *Performance Standard(s): At the beginning of each sprint, the Product Owner and development team will collaborate to define a set of user stories to be completed during the sprint. Acceptance criteria for each story will also be defined. The development team will deliver code and functionality to satisfy these user stories.*

> *Acceptable Quality Level: Delivered code meets the acceptance criteria for each user story. Incomplete stories will be assessed and considered for inclusion in the next sprint.*

- [ ] After 20 mins of inactivity, user gets notified that their sessions is about to expire. 
- [x] Notification contains options to extend or end the session
- [x] User will be able to log out using this interaction (end session)
- [x] Direct `Extend` button to `/v1/auth_check` endpoint
   - [x] Close alert box
- [x] Direct `Logout` button to `/v1/logout` endpoint
   - [x] Redirect to splash page with successful logout
- [x] User will be able to extend the session using this interaction
- [x] Session will automatically timeout at a specified length of time after 3 min notification
- [x] Testing Checklist has been run and all tests pass

**Note:** Times have been adjusted for the purpose of review and will be updated before merge. The testing times are 6 seconds of inactivity before the modal window appears and 10 seconds before automatic logout occurs

As Product Owner, @lfrohlich will decide if ACs are met.

### Deliverable 2: Tested Code

> *Performance Standard(s): Code delivered under the order must have substantial test code coverage. Version-controlled HHS GitHub repository of code that comprises products that will remain in the government domain.*

> *Acceptable Quality Level: Minimum of 90% test coverage of all code. All areas of code are meaningfully tested.*

+ Are all areas of code introduced in this PR meaningfully tested?
  + [ ] If this PR introduces backend code changes, are they meaningfully tested?
  + [ ] If this PR introduces frontend code changes, are they meaningfully tested?
+ Are code coverage minimums met?
  + [ ] Frontend coverage: ___ (see https://github.com/raft-tech/TANF-app README coverage dashboard)
  + [ ] Backend coverage: ___ (see https://github.com/raft-tech/TANF-app README coverage dashboard)

### Deliverable 3: Properly Styled Code

> *Performance Standard(s): GSA 18F Front- End Guide*

> *Acceptable Quality Level: 0 linting errors and 0 warnings*

+ [ ] Are backend code style checks passing on CircleCI?
+ [ ] Are frontend code style checks passing on CircleCI?
+ [ ] Does this PR change any linting or CI settings?

### Deliverable 4: Accessible

> *Performance Standard(s): Web Content Accessibility Guidelines 2.1 AA standards*

> *Acceptable Quality Level: 0 errors reported using an automated scanner and 0 errors reported in manual testing*

+ [ ] Did automated and manual testing with @iamjolly and @ttran-hub using Accessibility Insights reveal any errors introduced in this PR?
    + [See the full Accessibility Assessment plan here.](https://github.com/HHS/TANF-app/blob/main/docs/a11y/how-18f-will-test-a11y.md)

### Deliverable 5: Deployed

> *Performance Standard(s): Code must successfully build and deploy into the staging environment.*

> *Acceptable Quality Level: Successful build with a single command*

+ [ ] Was the code successfully deployed via automated CircleCI process to development on Cloud.gov?

### Deliverable 6: Documented

> *Performance Standard(s): Summary of user stories completed every two weeks. All dependencies are listed and the licenses are documented. Major functionality in the software/source code is documented, including system diagram. Individual methods are documented inline in a format that permits the use of tools such as JSDoc. All non-inherited 800-53 system security controls are documented in the Open Control or OSCAL format and HHS Section 508 Product Assessment Template (PAT) are updated as appropriate.*

> *Acceptable Quality Level: Combination of manual review and automated testing, if available*

+ [ ] If this PR introduces backend code, is that code documented both inline and overall?
+ [ ] If this PR introduces frontend code, is that code documented both inline and overall?
+ [ ] If this PR introduces dependencies, are their licenses documented?

### Deliverable 7: Secure

> *Performance Standard(s): Open Web Application Security Project (OWASP) Application Security Verification Standard 3.0*

> *Acceptable Quality Level: Code submitted must be free of medium- and high-level static and dynamic security vulnerabilities*

+ [ ] Does the OWASP Scan pass on CircleCI?
+ [ ] Do manual code review and manual testing detect any security issues?

## TO TEST

*List the steps to test the PR*

### This pull request is ready to merge when...

- [ ] Meets acceptance criteria for the issue 
- [ ] Code is reviewed by someone other than the original author
- [ ] Code is tested 
- [ ] Experience is approved by UX  
- [ ] Meets [a11y checklist](https://github.com/HHS/TANF-app/blob/main/docs/a11y/how-18f-will-test-a11y.md)
- [ ] Meets [Raft's Manual QASP Checklist](https://teams.microsoft.com/l/file/B5285301-13CA-46FD-A723-AFFD578C224A?tenantId=d58addea-5053-4a80-8499-ba4d944910df&fileType=docx&objectUrl=https%3A%2F%2Fhhsgov.sharepoint.com%2Fsites%2FTANFDataPortalOFA%2FShared%20Documents%2FGeneral%2FManual%20Testing%20Practices_Issue173.docx&baseUrl=https%3A%2F%2Fhhsgov.sharepoint.com%2Fsites%2FTANFDataPortalOFA&serviceName=teams&threadId=19:f769bbcb029f4f02b55ae7fad90e310d@thread.skype&groupId=41f194a6-c1d3-4680-933e-c8ee7d17e287):lock: (only applicable for PR to main HHS)
- [ ] Documentation is updated 
  - OpenAPI  
  - Readme
  - Entity Relationship Diagram (ERD)

### CODE REVIEW

File: tdrs-backend/tdpservice/settings/common.py
Comment (carltonsmith) [inline tdrs-backend/tdpservice/settings/common.py:45]: Using middleware to update the session so that it happens on every request (without having to add code to every request)

File: tdrs-backend/tdpservice/settings/common.py
Comment (carltonsmith) [inline tdrs-backend/tdpservice/settings/common.py:203]: This is the default, so if it isn't defined in the appropriate environment settings file it will be 30 minutes.

File: tdrs-backend/tdpservice/settings/development.py
Comment (carltonsmith) [inline tdrs-backend/tdpservice/settings/development.py:54]: This line will be removed before merge

File: tdrs-backend/tdpservice/settings/local.py
Comment (carltonsmith) [inline tdrs-backend/tdpservice/settings/local.py:27]: This line will be removed before merge

File: tdrs-backend/tdpservice/users/api/middleware.py
Comment (carltonsmith) [inline tdrs-backend/tdpservice/users/api/middleware.py:6]: This will update the HTTP Only cookie with a new expiration date/time on every request

File: tdrs-backend/tdpservice/users/api/utils.py
Comment (carltonsmith) [inline tdrs-backend/tdpservice/users/api/utils.py:158]: This is called on initial login

File: tdrs-backend/tdpservice/users/test/test_auth.py
Comment (carltonsmith) [inline tdrs-backend/tdpservice/users/test/test_auth.py:87]: This test ensures the cookie is updated with a new timestamp when a request is made to the API

File: tdrs-frontend/.env.development
Comment (hilvitzs) [inline tdrs-frontend/.env.development:4]: These are the environment variables for changing the timeout times for user sessions.

File: tdrs-frontend/package.json
Comment (hilvitzs) [inline tdrs-frontend/package.json:23]: This is the new package for dealing with session timeouts.

File: tdrs-frontend/src/assets/Button.scss
Comment (hilvitzs) [inline tdrs-frontend/src/assets/Button.scss:5]: I was individually styling multiple buttons the same way so I pulled that out into a single button style sheet.

File: tdrs-frontend/src/assets/EditProfile.scss
Comment (hilvitzs) [inline tdrs-frontend/src/assets/EditProfile.scss:12]: This was where I was previously doing button styling.

File: tdrs-frontend/src/assets/IdleTimer.scss
Comment (hilvitzs) [inline tdrs-frontend/src/assets/IdleTimer.scss:54]: This is styling for the session timeout modal.

File: tdrs-frontend/src/components/EditProfile/EditProfile.jsx
Comment (hilvitzs) [inline tdrs-frontend/src/components/EditProfile/EditProfile.jsx:208]: Removing unnecessary single button styling - replaced with global button styling

File: tdrs-frontend/src/components/Header/Header.jsx
Comment (hilvitzs) [inline tdrs-frontend/src/components/Header/Header.jsx:60]: In order for tabbing to work in navigation, I couldn't have the entirety of the `usa-nav__secondary` component be dynamically rendered. Instead, I changed the `Sign Out` button to change between `Sign In` and `Sign Out` depending on if the user is authenticated and only show the profile secondary link on authentication. This fixes a11y for keyboard navigation on the menu.

File: tdrs-frontend/src/components/Header/Header.jsx
Comment (hilvitzs) [inline tdrs-frontend/src/components/Header/Header.jsx:76]: This way the button will direct the user to the right place depending on their authentication.

File: tdrs-frontend/src/components/Header/Header.test.js
Comment (hilvitzs) [inline tdrs-frontend/src/components/Header/Header.test.js:131]: Adding tests for header component. Testing the new `Sign In` and `Sign Out` authentication change.

File: tdrs-frontend/src/components/IdleTimer/IdleTimer.jsx
Comment (hilvitzs) [inline tdrs-frontend/src/components/IdleTimer/IdleTimer.jsx:27]: This effect is for locking tabbing into the modal once it is open.

File: tdrs-frontend/src/components/IdleTimer/IdleTimer.jsx
Comment (hilvitzs) [inline tdrs-frontend/src/components/IdleTimer/IdleTimer.jsx:37]: This effect starts a secondary timer when the modal is opened so that a user can be logged out if they make no decision on the modal. This also focuses on the modal's `h1`.

File: tdrs-frontend/src/components/IdleTimer/IdleTimer.jsx
Comment (hilvitzs) [inline tdrs-frontend/src/components/IdleTimer/IdleTimer.jsx:69]: This is the function that ensures that the tabbing stays focused in the modal when it is open.

File: tdrs-frontend/src/components/IdleTimer/IdleTimer.jsx
Comment (hilvitzs) [inline tdrs-frontend/src/components/IdleTimer/IdleTimer.jsx:74]: This is for tabbing on the modal but also enables pressing escape to get out of the modal. Pressing escape will renew the session.

File: tdrs-frontend/src/components/IdleTimer/IdleTimer.jsx
Comment (hilvitzs) [inline tdrs-frontend/src/components/IdleTimer/IdleTimer.jsx:87]: This is the third party package that I'm using for the session timeout. The timeout variables come from the environment variables I defined. `Timeout` is how long before the modal pops up. `onAction` is called when there is an action and (since any action is sending a request to renew the auth token) those requests are debounced by the `debounce` key.

File: tdrs-frontend/src/components/IdleTimer/IdleTimer.test.js
Comment (hilvitzs) [inline tdrs-frontend/src/components/IdleTimer/IdleTimer.test.js:1]: Tests for the IdleTimer component.

File: tdrs-frontend/src/components/PrivateRoute/PrivateRoute.js
Comment (hilvitzs) [inline tdrs-frontend/src/components/PrivateRoute/PrivateRoute.js:43]: This is to ensure that the session timeout only runs on authenticated routes.

File: tdrs-frontend/src/components/Request/Request.jsx
Comment (hilvitzs) [inline tdrs-frontend/src/components/Request/Request.jsx:21]: I was able to remove the extra class name because the buttons are globally styled now.

File: tdrs-frontend/src/components/Header/Header.jsx
Comment (hilvitzs) [inline tdrs-frontend/src/components/Header/Header.jsx:69]: This is so that keyboard navigation works for navigation of the menus. With the way that the USWDS JS is written, it cannot dynamically render the secondary links and have them available for tabbing. At least one has to be there so that everything can be tabbed through. The profile `li` will be hidden when the user is not logged in.

File: tdrs-frontend/src/components/Header/Header.jsx
Comment (hilvitzs) [inline tdrs-frontend/src/components/Header/Header.jsx:88]: This is the `li` that will always be present. It will change whether it says `Sign In` or `Sign Out` depending on the authentication status.

**Comment by carltonsmith** [comment]
@alexsoble Snyk did not complete - I am hesitant to close and re-open to kick it off again because of all of the context we added

**Comment by iamjolly** [comment]
@hilvitzs - I'm seeing the timeout popup while I'm navigating the form with a screen reader. 

**Question for you:** As the idle timer taking into account changes in focused elements (like focus/blur changes happening in the form fields)?

**Something good to note:** When the modal appears, focus appropriately shifts to the modal's `h1` and the screen reader makes the expected announcement. 🎸 💥 💃

File: tdrs-backend/tdpservice/settings/common.py
Comment (alexsoble) [reply tdrs-backend/tdpservice/settings/common.py:203]: Look good! It might be helpful include the unit of measurement in the variable, something like `SESSION_TIMEOUT_MINUTES`, but that's a small suggestion and not something we have to wait on to merge this PR.

File: tdrs-backend/tdpservice/users/api/middleware.py
Comment (alexsoble) [inline tdrs-backend/tdpservice/users/api/middleware.py:31]: https://docs.djangoproject.com/en/3.1/ref/request-response/#django.http.HttpResponse.set_cookie

Setting `httponly=True` is a good security move. Should we set `samesite='Strict'`?

File: tdrs-frontend/.env.development
Comment (alexsoble) [reply tdrs-frontend/.env.development:4]: Gotcha! And looks like these are the testing values, and they'll be updated after merge, is that right?

**Comment by hilvitzs** [comment]
@iamjolly it should be updating the timer on any action on the page but let me look into that specifically. Is this with an actual screen reader or with voiceover? Or both?

File: tdrs-frontend/.env.development
Comment (hilvitzs) [reply tdrs-frontend/.env.development:4]: Exactly!

File: tdrs-frontend/src/components/IdleTimer/IdleTimer.jsx
Comment (alexsoble) [inline tdrs-frontend/src/components/IdleTimer/IdleTimer.jsx:8]: Inline documentation. 👍 

It might be nice to name these env variables and/or where a dev should look to see where they are set.

File: tdrs-frontend/src/components/IdleTimer/IdleTimer.test.js
Comment (alexsoble) [reply tdrs-frontend/src/components/IdleTimer/IdleTimer.test.js:1]: 👍 👍

**Comment by iamjolly** [comment]
> > @iamjolly it should be updating the timer on any action on the page but let me look into that specifically. Is this with an actual screen reader or with voiceover? Or both?
> 
> @hilvitzs
> 
> Both (I think). This would explain the issue I had with the original testing timeout speed (where I didn't have enough time on mobile to both switch focus and get the keyboard pulled up for a form field)

Yep... I'm using VoiceOver with Safari and while moving through the form, the modal interrupts. I can try to grab a quick video of it happening, if that will be helpful.

**Comment by hilvitzs** [comment]
> > > @iamjolly it should be updating the timer on any action on the page but let me look into that specifically. Is this with an actual screen reader or with voiceover? Or both?
> > 
> > 
> > @hilvitzs
> > Both (I think). This would explain the issue I had with the original testing timeout speed (where I didn't have enough time on mobile to both switch focus and get the keyboard pulled up for a form field)
> 
> Yep... I'm using VoiceOver with Safari and while moving through the form, the modal interrupts. I can try to grab a quick video of it happening, if that will be helpful.

Thank you but I should be able to reproduce it. I’ll let you know if I can’t for some reason. Thanks @reitermb for the info as well.

**Comment by hilvitzs** [comment]
@iamjolly I just looked on both local and the cloud and I am not getting that buggy behavior. The `onAction` method is getting triggered with a `debounce` of 1 second like expected. When I am tabbing through the form either with or without voiceover, the modal is not popping up as expected and when I stop for 6 seconds, the modal is popping up. If you could upload the video since I can't reproduce it, that would be great!

**Comment by alexsoble** [comment]
@iamjolly I tested this out in the cloud environment and I'm seeing the same behavior as @hilvitzs: tabbing resets the idle timer. (Testing using Chrome on MacBook.) 

You can observe this happening behind the scenes by opening the dev tools and watching the SET_AUTH and CLEAR_ALERT events.

**Comment by alexsoble** [comment]
# QASP Review 

### Deliverable 1: Accepted Features

> *Performance Standard(s): At the beginning of each sprint, the Product Owner and development team will collaborate to define a set of user stories to be completed during the sprint. Acceptance criteria for each story will also be defined. The development team will deliver code and functionality to satisfy these user stories.*

> *Acceptable Quality Level: Delivered code meets the acceptance criteria for each user story. Incomplete stories will be assessed and considered for inclusion in the next sprint.*

+ [ ] Look up the acceptance criteria in the related issue; paste ACs below in checklist format.
+ [ ] Check against the criteria:

As Product Owner, @lfrohlich will decide if ACs are met.

### Deliverable 2: Tested Code

> *Performance Standard(s): Code delivered under the order must have substantial test code coverage. Version-controlled HHS GitHub repository of code that comprises products that will remain in the government domain.*

> *Acceptable Quality Level: Minimum of 90% test coverage of all code. All areas of code are meaningfully tested.*

+ Are all areas of code introduced in this PR meaningfully tested?
  + [ ] If this PR introduces backend code changes, are they meaningfully tested?
  + [ ] If this PR introduces frontend code changes, are they meaningfully tested?
+ Are code coverage minimums met?
  + [ ] Frontend coverage: ___ (see https://github.com/raft-tech/TANF-app README coverage dashboard)
  + [ ] Backend coverage: ___ (see https://github.com/raft-tech/TANF-app README coverage dashboard)

### Deliverable 3: Properly Styled Code

> *Performance Standard(s): GSA 18F Front- End Guide*

> *Acceptable Quality Level: 0 linting errors and 0 warnings*

+ [ ] Are backend code style checks passing on CircleCI?
+ [ ] Are frontend code style checks passing on CircleCI?
+ [ ] Does this PR change any linting or CI settings?

### Deliverable 4: Accessible

> *Performance Standard(s): Web Content Accessibility Guidelines 2.1 AA standards*

> *Acceptable Quality Level: 0 errors reported using an automated scanner and 0 errors reported in manual testing*

+ [ ] Did automated and manual testing with @iamjolly and @ttran-hub using Accessibility Insights reveal any errors introduced in this PR?
    + [See the full Accessibility Assessment plan here.](https://github.com/HHS/TANF-app/blob/main/docs/a11y/how-18f-will-test-a11y.md)

### Deliverable 5: Deployed

> *Performance Standard(s): Code must successfully build and deploy into the staging environment.*

> *Acceptable Quality Level: Successful build with a single command*

+ [ ] Was the code successfully deployed via automated CircleCI process to development on Cloud.gov?

### Deliverable 6: Documented

> *Performance Standard(s): Summary of user stories completed every two weeks. All dependencies are listed and the licenses are documented. Major functionality in the software/source code is documented, including system diagram. Individual methods are documented inline in a format that permits the use of tools such as JSDoc. All non-inherited 800-53 system security controls are documented in the Open Control or OSCAL format and HHS Section 508 Product Assessment Template (PAT) are updated as appropriate.*

> *Acceptable Quality Level: Combination of manual review and automated testing, if available*

+ [ ] If this PR introduces backend code, is that code documented both inline and overall?
+ [ ] If this PR introduces frontend code, is that code documented both inline and overall?
+ [ ] If this PR introduces dependencies, are their licenses documented?

### Deliverable 7: Secure

> *Performance Standard(s): Open Web Application Security Project (OWASP) Application Security Verification Standard 3.0*

> *Acceptable Quality Level: Code submitted must be free of medium- and high-level static and dynamic security vulnerabilities*

+ [ ] Does the OWASP Scan pass on CircleCI?
+ [ ] Do manual code review and manual testing detect any security issues?

File: tdrs-frontend/src/components/IdleTimer/IdleTimer.jsx
Comment (hilvitzs) [reply tdrs-frontend/src/components/IdleTimer/IdleTimer.jsx:8]: I just pushed an update to the documentation in `IdleTimer` for where the environment variables are stored and what they're used for.

**Comment by iamjolly** [comment]
@hilvitzs and @alexsoble - thanks! I'll go in again this afternoon to re-test with VO+Safari and will make a recording if needed. What I'm wondering is: does the idle timer catch when a screen reader user is traversing the document without using `tab`? Some folks may use the VO+arrow keys to "read" through the page and form content/fields instead of using the `tab` key. That's why I asked about `focus` and `blur` earlier on. 

But, I'll re-test in a bit and share video if needed.

**Comment by hilvitzs** [comment]
> @hilvitzs and @alexsoble - thanks! I'll go in again this afternoon to re-test with VO+Safari and will make a recording if needed. What I'm wondering is: does the idle timer catch when a screen reader user is traversing the document without using `tab`? Some folks may use the VO+arrow keys to "read" through the page and form content/fields instead of using the `tab` key. That's why I asked about `focus` and `blur` earlier on.
> 
> But, I'll re-test in a bit and share video if needed.

Thank you very much! Here is a link to the package that I'm using for the idle timer: https://github.com/supremetechnopriest/react-idle-timer#default-events. These are the default events that trigger an activity: 
`mousemove`
`keydown`
`wheel`
`DOMMouseScroll`
`mousewheel`
`mousedown`
`touchstart`
`touchmove`
`MSPointerDown`
`MSPointerMove`
`visibilitychange`

so even though `blur` and `focus` aren't on there if they're using any key to traverse the form it should be triggered by `keydown`. If I'm misunderstanding something though then happy to fix it!

**Comment by iamjolly** [comment]
@hilvitzs - Here's a video I just made. I'm guessing that using the VoiceOver keys to read the page do not register as keypresses in the idle timer. It may not be a significant issue, though, with normal values used for timeouts in production when folks will likely press a key or activate a link/button to reset the timer during normal use. 

[TANF-idletimer-VoiceOver-probs.mov.zip](https://github.com/HHS/TANF-app/files/5592225/TANF-idletimer-VoiceOver-probs.mov.zip)

**Comment by reitermb** [comment]
Noting a mobile test I just did: It doesn't recognize iPadOS or Android screenreader focus changes either. I'm guessing that mobile screenreaders use something similar to the non-tab page reading @iamjolly mentioned?

**Comment by hilvitzs** [comment]
> @hilvitzs - Here's a video I just made. I'm guessing that using the VoiceOver keys to read the page do not register as keypresses in the idle timer. It may not be a significant issue, though, with normal values used for timeouts in production when folks will likely press a key or activate a link/button to reset the timer during normal use.
> 
> [TANF-idletimer-VoiceOver-probs.mov.zip](https://github.com/HHS/TANF-app/files/5592225/TANF-idletimer-VoiceOver-probs.mov.zip)

@iamjolly @alexsoble @carltonsmith Thank you so much Robert for the video, that is very helpful. I have never traversed a web page using that method so I'm glad you let me know. That being said, I can't figure out how exactly you are moving through elements using that method. I can get into the page but then I can't get anywhere else without pressing `tab`. Is it `ctrl` + `option` + `shift` + `down arrow`? And, if so, how do you move to other elements once you have moved into the web page?

Unfortunately, this package that I'm using does not allow me to add events to listen to (like `focus` or `blur`). I'm inclined to agree with your feeling about a user having another type of interaction (even if using a screen reader) sometime before the 20 minutes are up but if that ends up not being acceptable then we're most likely looking at replacing this package and either creating our own or finding a different package that allows for custom events.

**Comment by iamjolly** [comment]
> I can't figure out how exactly you are moving through elements using that method. I can get into the page but then I can't get anywhere else without pressing `tab`.

I'm using the [VoiceOver navigation command](https://www.apple.com/voiceover/info/guide/_1131.html#cpvokys04) (VO-right arrow and VO-left arrow) to move to the next/previous items to have it read the page. Otherwise tabbing just jumps from focusable item to focusable item and headings, paragraphs, and other non-focusable contents are skipped. 

> Unfortunately, this package that I'm using does not allow me to add events to listen to (like `focus` or `blur`). I'm inclined to agree with your feeling about a user having another type of interaction (even if using a screen reader) sometime before the 20 minutes are up but if that ends up not being acceptable then we're most likely looking at replacing this package and either creating our own or finding a different package that allows for custom events.

**I think you don't need to find a different package for this.** My sense is within 20 minutes, some other interaction that is detectable will occur on the page with screen reader users so this would be an extremely rare problem in production. These testing timeout values are pretty extreme, so I'm chalking that up to allowing us to catch this where the normal settings would never have revealed this.

I would, though, hope we can ask them to support other events like `scroll` or something customizable like detection of other focus changes so that additional accessibility coverage can be achieved with this in the future.

**Review by iamjolly — APPROVED** [review]
This looks good from an accessibility perspective based on the earlier comments and discussion around screen reader users and interaction idle time limits being artificially short in the testing environment. 

🚀  @alexsoble @lfrohlich @hilvitzs @carltonsmith 🚀

File: tdrs-backend/tdpservice/users/api/middleware.py
Comment (alexsoble) [reply tdrs-backend/tdpservice/users/api/middleware.py:31]: @carltonsmith Or would setting `samesite='Strict'` break things given that the frontend/backend are different apps? This also makes me think about our CSP settings.

**Pull request merged**
